src/reflect/scala/reflect/internal/util/Collections.scala

```diff
@@ -212,10 +212,23 @@ trait Collections {
   }
 
   final def mapFrom[A, A1 >: A, B](xs: List[A])(f: A => B): Map[A1, B] = {
-    Map[A1, B](xs map (x => (x, f(x))): _*)
+    val buf = immutable.Map.newBuilder[A1, B]
+    var ys = xs
+    while (!ys.isEmpty) {
+      val x = ys.head
+      buf += ((x, f(x)))
```

Comment:
Unfortunately, `MapBuilder` does not provide a `put` method, which would allow us to save us from creating a Tuple2 object in each iteration. Perhaps that could be added? On the other hand, I also need to check what is the efficiency of the default builder.

Reply:
This avoids tuples:

```scala
var m = immutable.Map.empty
while (...) {
  m = m.updated(k, v)
}
```

Reply:
Or using a combinator:

```
scala> def mapFrom[A, A1 >: A, B](xs: List[A])(f: A => B): Map[A1, B] = xs.foldLeft(collection.immutable.Map.empty[A1, B])((acc, a) => acc.updated(a, f(a)))
mapFrom: [A, A1 >: A, B](xs: List[A])(f: A => B)Map[A1,B]
```
(not tested)

Reply:
The problem is not just avoiding tuples, but also finding out what is the cost in allocations of the `updated` method from the immutable Map. Which is the default subclass used?

Reply:
Immutable map building is pretty inefficient in 2.12.x but [much improved](https://github.com/scala/scala/pull/7118) in 2.13.x.

The only potentially hot caller of `mapFrom` in the current compiler is `existentialBoundsExcludingHidden`, which is typically dealing with small maps. Probably not worth optimizing this area at the moment.

Reply:
(`MapBuilder` internally just delegates to `immutable.Map.+`, which is a `tuple` accepting version of `updated`)

Reply:
> Immutable map building is pretty inefficient in 2.12.x but much improved in 2.13.x.
>
> The only potentially hot caller of `mapFrom` in the current compiler is `existentialBoundsExcludingHidden`, which is typically dealing with small maps. Probably not worth optimizing this area at the moment.

so what's the verdict here? should this PR be closed for 2.12.x, and move on to the greener patches of 2.13.x+?

Reply:
@eed3si9n From Jason's comment, I will rebase and re-target to `2.13.x`